src/EditorFeatures/Core/Intents/IntentEditsSource.cs

```diff
@@ -0,0 +1,86 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Features.Intents;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Internal.Log;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor.Intents
+{
+    [Export(typeof(IIntentsEditsSource)), Shared]
+    internal class IntentEditsSource : IIntentsEditsSource
+    {
+        private readonly ImmutableDictionary<(string LanguageName, string IntentName), Lazy<IIntentProvider, IIntentProviderMetadata>> _lazyIntentProviders;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public IntentEditsSource([ImportMany] IEnumerable<Lazy<IIntentProvider, IIntentProviderMetadata>> lazyIntentProviders)
+        {
+            _lazyIntentProviders = CreateProviderMap(lazyIntentProviders);
+        }
+
+        private static ImmutableDictionary<(string LanguageName, string IntentName), Lazy<IIntentProvider, IIntentProviderMetadata>> CreateProviderMap(IEnumerable<Lazy<IIntentProvider, IIntentProviderMetadata>> lazyIntentProviders)
+        {
+            return lazyIntentProviders.ToImmutableDictionary(provider => (provider.Metadata.LanguageName, provider.Metadata.IntentName), provider => provider);
+        }
+
+        public async Task<IntentResult?> ComputeEditsAsync(IntentRequestContext intentRequestContext, CancellationToken cancellationToken)
+        {
+            var originalDocument = intentRequestContext.SnapshotBeforeIntent.GetOpenDocumentInCurrentContextWithChanges();
+            var currentDocument = intentRequestContext.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+
+            Contract.ThrowIfNull(originalDocument);
+            Contract.ThrowIfNull(currentDocument);
```

Comment:
ok for now.  likely somethign that should either bail gracefully on, or throw real ArgumentNullExceptions on.  this is a public entrypoint, so Contracts are not right.

Reply:
If we give something that is not correct, throw an exception like you do here. That will get our attention directly in the telemetry